### PR TITLE
Fix TaskGroupQueue will never be wakeup due to wakeup failed at one time

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Purpose of the pull request

- Fix the Task which using TaskGroup will never be wakeup due to wakeup failed at one time.

## Brief change log
In DS, we use TaskGroup to restrict the Task parallelism.
e.g. We have two task TaskA, TaskB, and they belong to the same task group, if we execute these two task at the same time, only one will get the taskGroup slot and the other will wait until be wakeup.
image
image
image

Unfortunately, if the TaskA wakeup TaskB failed, then TaskB will never be wakeup, since we use `put` mode, the waiting task will only be wakeup by other task, it will not try to pull the slot.

This PR will refactor this.
image

This PR add a new component `TaskGroupCoordinator` which use to deal with acquire/release task group slot and notify task.

Once a task need to acquire TaskGroup slot, it should call the `TaskGroupCoordinator::acquireTaskGroupSlot`, in this method, will insert a in queue TaskGroupQueue which status is WAITING.
Once a task is finished it will call `TaskGroupCoordinator::releaseTaskGroupSlot`, in this method, will move the TaskGroupQueue out queue.
image

And TaskGroupCoordinator will loop the slot, if there exist available slot, it will try to wakeup a waiting task.

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/repository/impl/TaskGroupDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    @Override
+    public List<TaskGroup> queryAvailableTaskGroups() {
+        return mybatisMapper.queryAvailableTaskGroups();
+    }
+
+    @Override
+    public boolean acquireTaskGroupSlot(Integer taskGroupId) {
+        if (taskGroupId == null) {
+            throw new IllegalArgumentException("taskGroupId cannot be null");
+        }
```
Comment (zhongjiajie) [inline dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/repository/impl/TaskGroupDaoImpl.java:59]: is there any reason we do not use `@NonNull` directly for parameter `taskGroupId`?

File: dolphinscheduler-dao/src/main/resources/sql/upgrade/3.3.0_schema/mysql/dolphinscheduler_ddl.sql [hunk 2]
```diff
@@ -80,4 +80,19 @@ ALTER TABLE `t_ds_process_definition_log` MODIFY COLUMN `version` int NOT NULL D
...
+BEGIN
+    DECLARE index_exists INT DEFAULT 0;
+    SELECT COUNT(*) INTO index_exists FROM information_schema.statistics WHERE table_schema = (SELECT DATABASE()) AND  table_name = 't_ds_task_group_queue' AND index_name = 'idx_t_ds_task_group_queue_in_queue';
+    IF index_exists = 0 THEN CREATE INDEX idx_t_ds_task_group_queue_in_queue ON t_ds_task_group_queue(in_queue);
+END IF;
+END;
+d//
+delimiter ;
+CALL create_idx_t_ds_task_group_queue_in_queue;
+DROP PROCEDURE create_idx_t_ds_task_group_queue_in_queue;
```
Comment (zhongjiajie) [inline dolphinscheduler-dao/src/main/resources/sql/upgrade/3.3.0_schema/mysql/dolphinscheduler_ddl.sql:98]: I think we could directly remove the related 3.3.0 upgrade. I would merge almost all upgrade sql from 3.3.0 to 3.2.1. and we will not release 3.3.0 until we close or nearly close 3.2.x

**Review by caishunfeng — APPROVED** [review]
LGTM

File: dolphinscheduler-dao/src/main/resources/sql/upgrade/3.3.0_schema/mysql/dolphinscheduler_ddl.sql (hunk 2 above)
Comment (ruanwenjun) [reply dolphinscheduler-dao/src/main/resources/sql/upgrade/3.3.0_schema/mysql/dolphinscheduler_ddl.sql:98]: Yes， we can remove the upgrade 3.3.0. We can use another PR to do this.

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/repository/impl/TaskGroupDaoImpl.java (hunk 1 above)
Comment (ruanwenjun) [reply dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/repository/impl/TaskGroupDaoImpl.java:59]: @NonNull Cannot write error message. If we want to use a specific error message ,we cannot use @NonNull.

**Comment by 2421994771** [comment]
I also encountered this problem, and it was found that when the in_queue field value was set to 1 in the wake-up task stage, the message was sent to the master failed due to network problems. Because the record with the value of IN_QUEUE was queried during the wake-up, the task would never be woken up in the subsequent period. Therefore, in the current production environment, my temporary solution was to query the records with 'status = -1 and in_queue = 1 and update_time < now() - 1 hour' in the t_ds_task_group_queue table associated with the workflow being executed through SQL, and update the in_queue value of these records to 0 so that they can be consumed again.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/taskgroup/TaskGroupCoordinator.java [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+                        taskGroupQueue);
+                releaseTaskGroupQueueSlot(taskGroupQueue);
+                continue;
+            }
+
+            if (taskInstance.getState().isFinished()) {
+                log.warn("The TaskInstance: {} state: {} finished, will release the TaskGroupQueue: {}",
+                        taskInstance.getName(), taskInstance.getState(), taskGroupQueue);
+                releaseTaskGroupQueueSlot(taskGroupQueue);
+                continue;
```
Comment (Gallardot) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/taskgroup/TaskGroupCoordinator.java:192]: Useless code

**Review by Gallardot — COMMENTED** [review]
Good job! Rest LGTM.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/taskgroup/TaskGroupCoordinator.java [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+                    // Notify the waiting task instance
+                    // We notify first, it notify failed, the taskGroupQueue will be in queue, and then we will retry it
+                    // next time.
+                    notifyWaitingTaskInstance(taskGroupQueue);
+
+                    // Set the taskGroupQueue status to RUNNING and remove from queue
+                    taskGroupQueue.setInQueue(Flag.YES.getCode());
+                    taskGroupQueue.setStatus(TaskGroupQueueStatus.ACQUIRE_SUCCESS);
+                    taskGroupQueue.setUpdateTime(new Date());
+                    taskGroupQueueDao.updateById(taskGroupQueue);
```
Comment (Gallardot) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/taskgroup/TaskGroupCoordinator.java:279]: There is some repetitive code in the state change that might be suitable for turning into a function.

Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/taskgroup/TaskGroupCoordinator.java:279]: Good suggestion, I will use another PR to move these kind of code in the whole project into function.

**Pull request merged**